### PR TITLE
JDK-8265042: javadoc HTML files not generated for types nested in records

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
This addresses an oversight in the support for records.

The fix could have been as simple as a one-word addition to a list of element kinds to check for nested classes. However, instead of that, the fix is more of a cleanup of a group of methods in `Utils` to fix this issue and to reduce the likelihood of anything similar happening again, when the next new kind of class is added.

The cleanup uses new code idioms using lambdas, predicates, and generic methods to replace varargs and temporary sets and copying lists.  The naming is somewhat cleaned up as well, but it was a self-imposed restriction that the cleanup is limited to the `Utils` class, and does not leak into clients of the class.

There is a core `getItems` method which is somewhat multi-purpose, supporting recursive and non-recursive use, and examining either just documented members or all members, including members that are not directly documented. (The latter occurs mostly in serialization.)   It is possible there is a future refactoring to further improve this method, perhaps to break it into different methods.  With that in mind, the signature of some methods was narrowed from accepting any `Element` to a specific kind of element: generally, one of `PackageElement` or `TypeElement`.

A new test is provided that checks that all the expected files are generated for various kinds of nested classes and interfaces in all kinds of classes and interfaces. The test is designed to fail if a new kind of class or interface is added in future without a corresponding update to the test.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265042](https://bugs.openjdk.java.net/browse/JDK-8265042): javadoc HTML files not generated for types nested in records

### Reviewers
 * [Hannes Wallnöfer](https://openjdk.java.net/census#hannesw) (@hns - **Reviewer**) ⚠️ Review applies to 23940d053ab838cbc601cff3a767914f0ad4df48

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3821/head:pull/3821` \
`$ git checkout pull/3821`

Update a local copy of the PR: \
`$ git checkout pull/3821` \
`$ git pull https://git.openjdk.java.net/jdk pull/3821/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3821`

View PR using the GUI difftool: \
`$ git pr show -t 3821`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3821.diff">https://git.openjdk.java.net/jdk/pull/3821.diff</a>

</details>

### CODE REVIEW

**Review by hns — APPROVED** [review]
I like both the predicate/lambda based methods and the new test a lot! I found some bits that should be cleaned up before integrating.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java [hunk 1]
```diff
@@ -2114,58 +2080,95 @@ public long getLineNumber(Element e) {
...
-            }
-        }
-        return elements;
+    /**
+     * Returns a list of immediately enclosed elements of a given class that match a predicate.
+     *
+     * @param e      the element, such as a package element or type element
+     * @param select the predicate for the selected members
+     * @param clazz  the class of the filtered members
+     * @param <T>    the class of the filtered members
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:2159]: Will fix the docs for `all`. `<T>` is supposed to be documented when it is in the signature of a generic method.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java [hunk 2]
```diff
@@ -2114,58 +2080,95 @@ public long getLineNumber(Element e) {
...
+     * @param filter the filter
+     * @param clazz  the class of the filtered members
+     * @param <T>    the class of the filtered members
+     */
+    private <T extends Element> void recursiveGetItems(Collection<T> list, Element e, boolean all, Predicate<Element> filter, Class<T> clazz) {
+        list.addAll(getItems0(e, all, filter, clazz));
+        List<TypeElement> classes = getItems0(e, all, this::isTypeElement, TypeElement.class);
+        for (TypeElement c : classes) {
+            list.addAll(getItems0(c, all, filter, clazz));
+            recursiveGetItems(list, c, all, filter, clazz);
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:2149]: The arguments for `getItems0` are different ...

[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:2150]: fixed

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java [hunk 3]
```diff
@@ -2114,58 +2080,95 @@ public long getLineNumber(Element e) {
...
+     * @param select the predicate for the selected members
+     * @param clazz  the class of the filtered members
+     * @param <T>    the class of the filtered members
+     *
+     * @return the list of enclosed elements
+     */
+    private <T extends Element> List<T> getItems0(Element e, boolean all, Predicate<Element> select, Class<T> clazz) {
+        return e.getEnclosedElements().stream()
+                .filter(e_ -> select.test(e_) && (all || shouldDocument(e_)))
+                .map(ee -> clazz.cast(ee))
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java]: done

**Pull request closed without merging**